### PR TITLE
New function to retrieve the version of JSON-RPC alongside the Rpc.ca…

### PR DESCRIPTION
**Pull request by kc284**
…ll from the request body.

Thus implementing clients can pass it to the other module functions requiring a
JSON-RPC version, in order to generate the right response.
Also, added some unit tests to cover various cases of request bodies.

Signed-off-by: Konstantina Chremmou <konstantina.chremmou@citrix.com>

### CODE REVIEW

File: tests/json.ml [hunk 1]
```diff
@@ -5,13 +5,122 @@ let good = "
  { \"a\":\"b\" }
 "
 
-let _ =
-  Printf.printf "Parsing good JSON ... %!";
-  let _ = Jsonrpc.of_string good in
+let v1 = "{
```
Comment (lindig) [inline tests/json.ml]: This can be simplified by using different string delimiters: 

```
{|
Can use " here
|}
```

File: tests/json.ml [hunk 2]
```diff
@@ -5,13 +5,122 @@ let good = "
...
+    let _ = Jsonrpc.call_of_string json in
+    Printf.printf "Passed: test %s\n" test_name;
+    true
+  with e ->
+    handle_exception test_name e pass
+
+let _ =
+  let results1 = tests1 |> List.map parse_json in
+  let results2 = tests2 |> List.map call_of_json in
+  let failures = results1 @ results2 |> List.filter (fun x -> not x) in
```
Comment (lindig) [inline tests/json.ml]:
```
match List.concat [results1; results2] |> List.filter not with
| [] -> ()
| xs -> failwith (Printf.sprintf "Test failures: %d"  (List.length xs))
```

It might be worth using a proper test framework because the result does not tell which tests fail.

File: tests/json.ml (hunk 1 above)
Comment (kc284) [inline tests/json.ml]: For some reason I can't get it to compile 🤔

File: tests/json.ml (hunk 2 above)
Comment (kc284) [inline tests/json.ml]: All the results (success and failure) are printed out before we get to this line, but I agree, it would be better to use OUnit. However, would it be ok if done as part of a different PR, because (a) there are several more tests in this repo to be transformed, and (b) I need this change in as soon as possible because it blocks a different task I'm in the middle of 😟

File: tests/json.ml (hunk 1 above)
Comment (mseri) [inline tests/json.ml]: I like @lindig suggestion but I think this can be in the scope for a different PR where we update the whole test. As annoying to read and write as it is, it is anyway following the style of the other strings

File: tests/json.ml (hunk 2 above)
Comment (mseri) [inline tests/json.ml]: Using OUnit and QCheck here is worth, but I agree it is a lot of work. Please make the PR updating the tests in the future if you can. If @lindig is fine with it we may postpone this

**Comment by kc284** [comment]
One thing that's more to the point, and I considered it today, is whether more checks should be added around line 607 to check the internal structure of the error.

File: tests/json.ml (hunk 1 above)
Comment (lindig) [inline tests/json.ml]: I suspect some interaction between camlp4 and `{| |}` strings. I don't understand why they are not working here. So keep the `"..."` strings.

File: lib/jsonrpc.ml [hunk 3]
```diff
@@ -554,27 +554,41 @@ let get name dict =
...
-let call_of_string str =
+let version_and_call_of_string str =
   match of_string str with
   | Dict d ->
-    let name = match get "method" d with String s -> s | _ -> raise (Malformed_method_request str) in
-    let params =
+    let name =
+      match get "method" d with
+      | String s -> s
+      | _ -> raise (Malformed_method_request "method")
```
Comment (mseri) [inline lib/jsonrpc.ml]: I like specified what failed, but I think having also the rest of the string in here would make it easier to debug.
What about adding:
```ocaml
let malformed_method_error loc = Malformed_method request (Printf.sprintf "\"%s\" in %s" loc str) in 
```
and then `raise (malformed_method_error "method")` here and similarly in the calls below

File: tests/json.ml [hunk 4]
```diff
@@ -5,13 +5,122 @@ let good = "
...
+}"
+
+let tests1 = [
+  "bad", bad, false;
+  "good", good, true;
+]
+
+let tests2 = [
+  "v1", v1, true;
+  "v1_null_id", v1_null_id, false; (*stricter than the specs*)
```
Comment (mseri) [inline tests/json.ml]: It would be worth fix the `stricter` ones in a forthcoming PR

**Comment by mseri** [comment]
Concerning the internal structure error, I think we should make a backward compatible change adding an optional parameter, e.g. `?(strict:false), that allows to enable strict checking on the structure of the error payload and fails if it is not standard compliant.

File: lib/jsonrpc.ml (hunk 3 above)
Comment (kc284) [inline lib/jsonrpc.ml]: I'd rather modify the message if the Malformed_method_request to "method" plus the str, because changing the type of the exception thrown affects the interface.

File: tests/json.ml (hunk 4 above)
Comment (kc284) [inline tests/json.ml]: I considered changing them in this PR (the sooner the better), but @robhoes recommended to leave it as is. To my understanding we don't have a use case for them.

File: lib/jsonrpc.ml [hunk 5]
```diff
@@ -554,28 +558,42 @@ let get name dict =
...
     let (_:int64) = match get "id" d with Int i -> i | _ -> raise (Malformed_method_request str) in
-    call name params
+    version, call name params
   | _ -> raise (Malformed_method_request str)
 
-let response_of_stream str =
+let call_of_string str =
+  snd ( version_and_call_of_string str)
+
+let response_of_stream ?(strict=false) str =
```
Comment (robhoes) [inline lib/jsonrpc.ml]: I think that we can get away without no having this `strict` argument, and always being strict. We are just following the spec here, so clients need to do it properly.

File: lib/jsonrpc.ml (hunk 3 above)
Comment (jonludlam) [inline lib/jsonrpc.ml]: You've already renamed the function, so there's no problem in changing the exception raised also. You can always catch and rethrow in your compatibility function below.

Comment (kc284) [inline lib/jsonrpc.ml]: @jonludlam I haven't renamed the function, I only changed its internals and added a new function. If someone already uses the old function and upgrades, their code will still compile, but if they're catching Malformed_method_request and the method suddenly starts throwing malformed_method_error it will be an issue. Or am I missing something?

Comment (robhoes) [inline lib/jsonrpc.ml]: `call_of_rpc` was renamed to `version_and_call_of_rpc`. And then a "new" `call_of_rpc` was added for compat, which calls `version_and_call_of_rpc`. So you can catch a new exception from `version_and_call_of_rpc` in `call_of_rpc` and reraise as the old one.

Comment (kc284) [inline lib/jsonrpc.ml]: Ah ok, I see what you mean, then the interface is the same.
However, what is the benefit of doing this instead of raising the original exception with extended string (to also avoid the additional overhead of catch-rethrow)?

Comment (jonludlam) [inline lib/jsonrpc.ml]: I was going to say:
```
It allows us to throw a better error - while it's nice to know what went wrong (e.g. "method") it'd also be nice to see the string we were trying to parse.
```
but I now realise we're passing the string in here, so it adds nothing to return it in the exception. In which case, throwing the exception with the thing that went wrong ("method") seems sensible to me.

However, since you're calling `get` rather than `get'`, you're going to get a different type of exception depending on whether `get` fails or returns you something of the wrong type. That should be fixed.

File: lib/jsonrpc.ml [hunk 6]
```diff
@@ -607,7 +611,14 @@ let response_of_stream str =
       let error = get' "error" d in
       begin match result, error with
         | Some v, None    -> success v
-        | None, Some v    -> failure v
```
Comment (mseri) [inline lib/jsonrpc.ml]: So now we are always checking?

**Comment by kc284** [comment]
Very likely, however it is better to be compliant with the specs, which most people would expect, than having an implementation working with a restricted number of clients. 
We may want to consider bumping the major version.

**Comment by mseri** [comment]
As discussed in today's meeting I am going to merge and release `2.0.0`
